File: backend/internal/domain/banner/banner_config.go
package banner

// Config holds the configuration for different banner types.
type Config struct {
	BaseRate         float64
	FourStarRate     float64
	SoftPityStart    int
	HardPity         int
	RateIncrease     float64
	RateUpChance     float64
	GuaranteedRateUp bool
}

// bannerConfigs stores the configuration for all banner types.
var bannerConfigs = map[Type]Config{
	StarRailLimited: {
		BaseRate:         0.006, // 0.6%
		FourStarRate:     0.051,
		SoftPityStart:    73,
		HardPity:         90,
		RateIncrease:     0.07,
		RateUpChance:     0.5,
		GuaranteedRateUp: true,
	},
	StarRailLightCone: {
		BaseRate:         0.008, // 0.8%
		FourStarRate:     0.066,
		SoftPityStart:    65,
		HardPity:         80,
		RateIncrease:     0.07,
		RateUpChance:     0.75,
		GuaranteedRateUp: true,
	},
	GenshinLimited: {
		BaseRate:         0.006, // 0.6%
		FourStarRate:     0.051,
		SoftPityStart:    73,
		HardPity:         90,
		RateIncrease:     0.07,
		RateUpChance:     0.5,
		GuaranteedRateUp: true,
	},
	GenshinWeapon: {
		BaseRate:         0.007, // 0.7%
		FourStarRate:     0.066,
		SoftPityStart:    62,
		HardPity:         80,
		RateIncrease:     0.07,
		RateUpChance:     0.75,
		GuaranteedRateUp: true,
	},
	GenshinStandard: {
		BaseRate:         0.006, // 0.6%
		FourStarRate:     0.051,
		SoftPityStart:    73,
		HardPity:         90,
		RateIncrease:     0.07,
		RateUpChance:     0.5,
		GuaranteedRateUp: false,
	},
	ZenlessLimited: {
		BaseRate:         0.006, // 0.6%
		FourStarRate:     0.051,
		SoftPityStart:    73,
		HardPity:         90,
		RateIncrease:     0.07,
		RateUpChance:     0.5,
		GuaranteedRateUp: true,
	},
	ZenlessWEngine: {
		BaseRate:         0.01, // 1%
		FourStarRate:     0.08,
		SoftPityStart:    64,
		HardPity:         80,
		RateIncrease:     0.07,
		RateUpChance:     0.75,
		GuaranteedRateUp: true,
	},
	ZenlessStandard: {
		BaseRate:         0.006, // 0.6%
		FourStarRate:     0.051,
		SoftPityStart:    73,
		HardPity:         90,
		RateIncrease:     0.07,
		RateUpChance:     0.5,
		GuaranteedRateUp: false,
	},
	ZenlessBangboo: {
		BaseRate:         0.01, // 1%
		FourStarRate:     0.051,
		SoftPityStart:    64,
		HardPity:         80,
		RateIncrease:     0.07,
		RateUpChance:     1.0,
		GuaranteedRateUp: true,
	},
	StarRailStandard: {
		BaseRate:         0.006, // 0.6%
		FourStarRate:     0.051,
		SoftPityStart:    73,
		HardPity:         90,
		RateIncrease:     0.07,
		RateUpChance:     0.5,
		GuaranteedRateUp: false,
	},
}

// GetConfig returns the configuration for a specific banner type.
func GetConfig(bannerType Type) Config {
	if config, ok := bannerConfigs[bannerType]; ok {
		return config
	}
	// Default to StarRailStandard if unknown type
	return bannerConfigs[StarRailStandard]
}

File: stats/core/banner_config.py
"""Configuration for different banner types."""

from core.banner import BannerConfig

BANNER_CONFIGS = {
    "star_rail_standard": BannerConfig(
        game_name="Star Rail",  # Added
        banner_type="Standard",  # Added
        base_rate=0.006,
        four_star_rate=0.051,
        soft_pity_start_after=73,
        hard_pity=90,
        rate_increase=0.07,
        guaranteed_rate_up=False,
        rate_up_chance=0.5,
    ),
    "star_rail_limited": BannerConfig(
        game_name="Star Rail",
        banner_type="Limited",  # Updated
        base_rate=0.006,
        four_star_rate=0.051,
        soft_pity_start_after=73,
        hard_pity=90,
        rate_increase=0.07,
        guaranteed_rate_up=True,
        rate_up_chance=0.5,
    ),
    "star_rail_light_cone": BannerConfig(
        game_name="Star Rail",  # Added
        banner_type="Light Cone",  # Added
        base_rate=0.008,
        four_star_rate=0.066,
        soft_pity_start_after=65,
        hard_pity=80,
        rate_increase=0.07,
        guaranteed_rate_up=True,
        rate_up_chance=0.75,
    ),
    "genshin_standard": BannerConfig(
        game_name="Genshin Impact",  # Added
        banner_type="Standard",  # Added
        base_rate=0.006,
        four_star_rate=0.051,
        soft_pity_start_after=73,
        hard_pity=90,
        rate_increase=0.07,
        guaranteed_rate_up=False,
        rate_up_chance=0.5,
    ),
    "genshin_limited": BannerConfig(
        game_name="Genshin Impact",
        banner_type="Limited",  # Updated
        base_rate=0.006,
        four_star_rate=0.051,
        soft_pity_start_after=73,
        hard_pity=90,
        rate_increase=0.07,
        guaranteed_rate_up=True,
        rate_up_chance=0.5,
    ),
    "genshin_weapon": BannerConfig(
        game_name="Genshin Impact",  # Added
        banner_type="Weapon",  # Added
        base_rate=0.007,
        four_star_rate=0.066,
        soft_pity_start_after=62,
        hard_pity=80,
        rate_increase=0.07,
        guaranteed_rate_up=True,
        rate_up_chance=0.75,
    ),
    "zenless_standard": BannerConfig(
        game_name="Zenless Zone Zero",  # Added
        banner_type="Standard",  # Added
        base_rate=0.006,
        four_star_rate=0.051,
        soft_pity_start_after=73,
        hard_pity=90,
        rate_increase=0.07,
        guaranteed_rate_up=False,
        rate_up_chance=0.5,
    ),
    "zenless_limited": BannerConfig(
        game_name="Zenless Zone Zero",
        banner_type="Limited",  # Updated
        base_rate=0.006,
        four_star_rate=0.051,
        soft_pity_start_after=73,
        hard_pity=90,
        rate_increase=0.07,
        guaranteed_rate_up=True,
        rate_up_chance=0.5,
    ),
    "zenless_w_engine": BannerConfig(
        game_name="Zenless Zone Zero",  # Added
        banner_type="W-Engine",  # Added
        base_rate=0.01,
        four_star_rate=0.08,
        soft_pity_start_after=64,
        hard_pity=80,
        rate_increase=0.07,
        guaranteed_rate_up=True,
        rate_up_chance=0.75,
    ),
    "zenless_bangboo": BannerConfig(
        game_name="Zenless Zone Zero",  # Added
        banner_type="Bangboo",  # Added
        base_rate=0.01,
        four_star_rate=0.051,
        soft_pity_start_after=64,
        hard_pity=80,
        rate_increase=0.07,
        guaranteed_rate_up=True,
        rate_up_chance=1.0,
    ),
}
